File: samples/src/main/java/com/mobilesolutionworks/android/controller/samples/ui/databinding/DataBinding.java
package com.mobilesolutionworks.android.controller.samples.ui.databinding;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.lang.ref.WeakReference;

/**
 * Created by yunarta on 17/3/17.
 */

public class DataBinding {

    public static class ViewHolder<U, V extends ViewDataBinding> extends RecyclerView.ViewHolder {

        private final V mBinding;

        private U mUserObject;

        public ViewHolder(V binding) {
            super(binding.getRoot());
            mBinding = binding;
        }

        public final V getBinding() {
            return mBinding;
        }

        public final U getUserObject() {
            return mUserObject;
        }

        public final void setUserObject(U userObject) {
            mUserObject = userObject;
        }
    }

    public static abstract class Adapter<U> extends RecyclerView.Adapter<ViewHolder<U, ? extends ViewDataBinding>> {

        private WeakReference<LayoutInflater> mInflater;

        protected abstract int getItemLayout(int viewType);

        private LayoutInflater createInflaterIfNeeded(ViewGroup parent) {
            if (mInflater == null || mInflater.get() == null) {
                mInflater = new WeakReference<>(LayoutInflater.from(parent.getContext()));
            }

            return mInflater.get();
        }

        @Override
        public ViewHolder<U, ? extends ViewDataBinding> onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ViewHolder<>(DataBindingUtil.inflate(createInflaterIfNeeded(parent), getItemLayout(viewType), parent, false));
        }
    }

    public static abstract class SingleTypeAdapter<U, V extends ViewDataBinding> extends RecyclerView.Adapter<ViewHolder<U, V>> {

        private WeakReference<LayoutInflater> mInflater;

        protected abstract int getItemLayout();

        private LayoutInflater createInflaterIfNeeded(ViewGroup parent) {
            if (mInflater == null || mInflater.get() == null) {
                mInflater = new WeakReference<>(LayoutInflater.from(parent.getContext()));
            }

            return mInflater.get();
        }

        @Override
        public ViewHolder<U, V> onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ViewHolder<>(DataBindingUtil.inflate(createInflaterIfNeeded(parent), getItemLayout(), parent, false));
        }
    }
}

File: example-app/src/main/java/com/mobilesolutionworks/android/controller/samples/ui/databinding/DataBinding.kt
package com.mobilesolutionworks.android.controller.samples.ui.databinding

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import java.lang.ref.WeakReference

/**
 * Created by yunarta on 17/3/17.
 */

interface DataBinding {

    class ViewHolder<U, V : ViewDataBinding>(val binding: V) : RecyclerView.ViewHolder(binding.root) {

        var userObject: U? = null
    }

    abstract class Adapter<U> : RecyclerView.Adapter<ViewHolder<U, out ViewDataBinding>>() {

        private var mInflater: WeakReference<LayoutInflater>? = null

        protected abstract fun getItemLayout(viewType: Int): Int

        private fun createInflaterIfNeeded(parent: ViewGroup): LayoutInflater {
            if (mInflater == null || mInflater!!.get() == null) {
                mInflater = WeakReference(LayoutInflater.from(parent.context))
            }

            return mInflater!!.get()!!
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<U, out ViewDataBinding> {
            return ViewHolder(DataBindingUtil.inflate<ViewDataBinding>(createInflaterIfNeeded(parent), getItemLayout(viewType), parent, false))
        }
    }

    abstract class SingleTypeAdapter<U, V : ViewDataBinding> : RecyclerView.Adapter<ViewHolder<U, V>>() {

        private var mInflater: WeakReference<LayoutInflater>? = null

        protected abstract fun getItemLayout(): Int

        private fun createInflaterIfNeeded(parent: ViewGroup): LayoutInflater {
            if (mInflater == null || mInflater!!.get() == null) {
                mInflater = WeakReference(LayoutInflater.from(parent.context))
            }

            return mInflater!!.get()!!
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<U, V> {
            return ViewHolder(DataBindingUtil.inflate<V>(createInflaterIfNeeded(parent), getItemLayout(), parent, false))
        }
    }
}